### ISSUE TITLE
Move grid information off individual layers

### ISSUE THREAD

**Issue by sofroniewn** [napari/napari#1690]
## Refactor

As part of improvements to our handling of transforms I propose we move `grid` mode off of the individual layers and have all of that done at the level of the viewer

**Comment by GenevieveBuckley** [comment]
This sounds like a good idea. 

What impact do you think this change will have on the current situation? Eg: right now when I select points the coordinates are very different depending on whether grid mode is on or off, so I can't toggle between these views.

One situation I currently find frustrating is that we don't have an easy way to do matched control point selections. We avoided doing this in the [sudoku example](https://github.com/napari/napari/pull/1616#issuecomment-700146465) by using RANSAC, but for a lot of real-world applications I really want to be able to have two images side by side while I pick out the matched control points. Since example use cases are helpful, here's [a non-napari GUI I made for our lab](https://github.com/DeMarcoLab/correlateim), which uses [this python implementation](https://github.com/hofmann-tobias/cpselect) of the MATLAB cpselect function.

**Comment by sofroniewn** [comment]
> What impact do you think this change will have on the current situation? Eg: right now when I select points the coordinates are very different depending on whether grid mode is on or off, so I can't toggle between these views.
>
> One situation I currently find frustrating is that we don't have an easy way to do matched control point selections. 

We could try and get this fixed!! It might be a little tricky without having two canvases, but we can try. Thanks for the links. I do want us to push on the multicanvas too once all of this is done.

From an implementation standpoint, we'd probably need to pass the grid transform to the [base vispy visual](https://github.com/napari/napari/blob/master/napari/_vispy/vispy_base_layer.py) somehow and remove it from the [base layer itself](https://github.com/napari/napari/blob/master/napari/layers/base/base.py#L184-L186).

We could also try pulling all the [grid code out of the viewer_model](https://github.com/napari/napari/blob/master/napari/components/viewer_model.py#L404-L458) and into a little object called `Grid`, so you could do things like `viewer.grid.stride` and `viewer.grid.n_columns` etc.

One big question is then how to handle the camera - that now exists in a gridworld space, but the cursor and be made to work in the world space of the selected layer, irrespective of where in the grid it is. You should then be able to mark up images as you want with points in the correct coordinate system irrespective of where things are in the grid. Am I making sense here or do you think I have it way off? It's definitely tricky.

If you want to start working on this I'm happy to review PRs/ try and offer guidance

**Comment by GenevieveBuckley** [comment]
Muticanvas support sounds like it would help do what we want - probably it's better to wait and implement that properly, rather than messing around with stuff here before that happens?

**Comment by GenevieveBuckley** [comment]
> One big question is then how to handle the camera - that now exists in a gridworld space, but the cursor and be made to work in the world space of the selected layer, irrespective of where in the grid it is. You should then be able to mark up images as you want with points in the correct coordinate system irrespective of where things are in the grid. Am I making sense here or do you think I have it way off? It's definitely tricky.

I think you're making sense, but there are some problems. Because points are layers themselves, we can't work with them in grid mode if they are supposed to indicate points on an Image layer.

I've told Juan recently that I'm not a fan of doing tricky things with strides in grid mode to force the outcome I want. It's much too fragile, if a layer is added, removed, or even rearranged then it will break the whole carefully constructed house of cards.

**Comment by sofroniewn** [comment]
> Muticanvas support sounds like it would help do what we want - probably it's better to wait and implement that properly, rather than messing around with stuff here before that happens?
>
> I've told Juan recently that I'm not a fan of doing tricky things with strides in grid mode to force the outcome I want. It's much too fragile, if a layer is added, removed, or even rearranged then it will break the whole carefully constructed house of cards.

yeah, those are both fair points. we might be best waiting a little longer, but it's definitely good to keep thinking about this

**Comment by brisvag** [comment]
This was actually closed a while back by #7870.
